### DISCUSSION TITLE
Improvements to Duplication of Records

### DISCUSSION THREAD

**Discussion by emilyf** [directus/directus#16112] [Feature Requests]
I've been rolling out the admin UI to some of my users and there's a lot of feedback around how duplication of records work. Some thoughts to start the discussion (I hope this isn't a dupe and apologies, I searched!)

1. Right now to even do a save as copy you have to modify something on an existing record to get the three dots:
<img width="1098" alt="Screen Shot 2022-10-21 at 2 25 54 PM" src="https://user-images.githubusercontent.com/1789807/197282879-5383c417-6f82-4279-a9d0-1ae4c04fb7e9.png">

It seems a more standard UX approach to this would be to have the well known duplicate icon visible when you go into a record before modifying anything. 
<img width="55" alt="Screen Shot 2022-10-21 at 2 26 44 PM" src="https://user-images.githubusercontent.com/1789807/197282987-3129b6be-09ef-488a-9688-fb717328f8ff.png">

2. Upon clicking icon (or if still using same three dots) there's no real "obvious" way to show you that you are now editing the duplicate record. I was thinking one of the two below approaches:
- add in brackets a generally known UX word [Copy] into title field
- Show the user a message -- this could be a notification pop up but to be even more obvious maybe a message at top of screen using same class/appearance as below that says '$Collection' item duplicated, continue entering information below' (best I could come up with, sure there's a shorter version)....
<img width="311" alt="Screen Shot 2022-10-21 at 2 31 05 PM" src="https://user-images.githubusercontent.com/1789807/197283600-404d1f13-2041-4e16-955b-21c04ad305a4.png">

I think it's much more of a normal experience to NOT have to first edit a field to make a duplicate, so I would love to go with something closer to the first idea ....

Thoughts?

**Comment by connorwinston** [comment]
I totally agree, this is something that throws me off sometimes when duplicating records. Building off your thoughts...

> It seems a more standard UX approach to this would be to have the well known duplicate icon visible when you go into a record before modifying anything. 

**1. Maybe just changing `Save As Copy` to `Duplicate` and changing the behavior.**

> 2. Upon clicking icon (or if still using same three dots) there's no real "obvious" way to show you that you are now editing the duplicate record. I was thinking one of the two below approaches:
> 
> * add in brackets a generally known UX word [Copy] into title field
> * Show the user a message -- this could be a notification pop up but to be even more obvious maybe a message at top of screen using same class/appearance as below that says '$Collection' item duplicated, continue entering information below' (best I could come up with, sure there's a shorter version)....

**2. We could just re-use the `Creating Item in Collection` "design" and make it `Creating Copy in Collection`.** 

> I think it's much more of a normal experience to NOT have to first edit a field to make a duplicate, so I would love to go with something closer to the first idea ....

**3. I agree, whenever I think of Duplicate I think of it taking me to a new create form that is prefilled with the fields marked as allowing duplicates. However, the current `Save as Copy` terminology matches the current behavior where it requires edits and saves the current item/edits as a copy.**

We shall have to see what the rest of the team thinks and @benhaynes since he does most of the design related ideas.

**Reply by benhaynes** [reply]
From a design perspective, I think the best approach here is to update the "..." icon that is currently only for alternate saving... to be for any/all extended options for the header. It would always be visible, might be a different icon (or not), and could have all sorts of options within. That way, the options _within_ the dropdown could be individually enabled/disabled as needed.

Then we could have all sorts of options in there, including `Duplicate` or `Make a Copy`, `Full-Screen Mode`, `Print`, `Save as PDF`, `Share` (just opens share drawer), etc.

I'm seeing this as a sort of "File Menu" for the item, or an overflow for Action Buttons (based on screen real estate). This also would help with how to handle all those header action buttons on mobile in a more elegant way. In the future, we could allow configuring which of these dropdown options are shown as dedicated Action buttons (vs just in the actions dropdown).

Does that ramble make sense?

**Reply by emilyf** [reply]
So if I'm understanding, this approach would change the "..." icon to something more intuitive and be for everything in header area (except I assume the checkbox and delete). And as you mention have it always present and not grayed out. So my answer I think is a hesitant yes and a hesitant no. A lot of it is absurdly based on number of clicks, so expanding is always that extra click (I DO think this is an awesome idea, my ONLY hesitancy is if the duplicate option goes in there or gets its own icon). 

It's such a small thing but I really do get endless feedback on this from my users. 

Let me try to explain with a walk through real world example. I do think this meets 80/20 rule as the use case applies to any job where a person's responsibilities include something related to rapid duplication -- often as placeholder content that relates to deadlines or scheduling -- think local journalism outlet prescheduling local high school athlete "highlight" for every friday, managing social media content calendar for a company, creating / updating inventory when you have to log every chair in the building individually, booking/reserving a room in a library or being a local video producer who rents the same equipment package every week at the same time. Tons of situations where a lot of repetition comes into play, but here is a very specific example. If TLDR just skip to "Overall Conclusion" at very bottom of this comment or here is my TLDR summary:

**Using the drawer idea you suggested, I think if the record could be left unedited and there was a shortcode for "save as copy" then it would be fine to keep it in the drawer, and giving an in-app notification you're now editing the duplicate record would probably solve this**

**Use Case** is to explain need for suggested solution above -- Note I'm inserting myself into this use case but all of the steps and confusion I include come from direct feedback from the users on our platform, and the length is applicable based on the info users provide in tickets.

Let's say I work for ABC Network as the content scheduler. All of our metadata and video file locations etc are stored in Directus. My job is to create the live schedule for the entire quarter and get sign off. Since an hour primetime show standard duration is 47 minutes and 1/2 hour is 22 minutes, I just need a ton of placeholders that eventually get updated with the real metadata as the quarter progresses. I'll use Grey's Anatomy considering it's somehow managed to create 403 episodes through 19 seasons (no judgement here, just stating the facts); obviously, it's going on the primetime so I start with that. As the scheduler,  I need individual instances in the schedule because they will eventually be connected to a video file; but the program name is always the same, duration, producers --all the metadata is the same except the episode title, filename it connects to, and episode description. Updating the description and filename is not my job, someone else does that, I just put in the placeholders for its air times for the whole quarter so I need to duplicate and only modify the episode title (aka Episode 404, Episode 405, etc for whole quarter). 

This is pretty mundane work, and I've got to do it for every other show in the lineup for the entire quarter; I'm probably trying to crank it out in a couple days, get feedback, make changes, and eventually get sign off. 

In any workflow like this, number of clicks is king; I just want to work as fast as possible, that's number 1 priority. Second priority is that it's second nature/easy because I'm most likely listening to a podcast at the same time as it's just repetition. Let's also assume I just got hired and have never used Directus, but that I pick things up reasonably quick when it comes to my day to day software. 

So keeping this in mind, the current individual item UI, even with the above suggestions, does not meet my two top priorities (# of clicks and second nature). Since I didn't get trained in Directus cause my manager left with no notice and the IT person only works one day a week and we just switched to this new platform before I came on, I'm on my own. So I sit down to log in for the first time and get ready to roll. I go into the last episode from Grey's Anatomy and see

<img width="1096" alt="Screen Shot 2022-10-26 at 11 54 58 AM" src="https://user-images.githubusercontent.com/1789807/198100660-bd2e2ded-cd75-4dda-ae7b-a9f8d59e1982.png">


I don't want to change the _current_ episode I'm looking at, I want to duplicate it, placehold it, repeat for the entire lineup in order to fill 24/7 hours of airtime for an entire quarter. 

So I'm  looking for the "copy/duplicate" button or icon. (Note: this is especially true for very new users to directus, but even people who have been using it for a little bit for some reason seem to still get stuck on this -- see my final thoughts on why at bottom).

I can't find it. The top save button is grayed out, i see the trash icon and understand what these mean but that just seems like the place where that duplicate option should be. Oh well, it's probably at the bottom so my next instinct is to scroll down thinking it must be down there; it's not, so I scroll back up to the top thinking I for sure just missed it. Still don't see it. (Have to pause my podcast at this point). I start looking on the right side and see "revisions", open that up, no dice, click the other areas, can't find it. 

The most common next step my users take, and thus I'm going to do it in my example, is I go _back_ to the collection list assuming I must have to check off an item and then there will be a "duplicate" option somewhere on that page. I don't even think to click in a field on the item I was looking at because again, I don't want to change that record. So after I can't find a way on the collection list screen I go back into Episode 403 cause it's gotta be there. Can't find it, so I start clicking fields -- however, I'm just clicking not changing anything because I don't want anything changed, so I still have a grayed out check button and this duplication feature is starting to be a head scratcher. Finally I just click into the title field and type some jibberish because my other option is to report it as a bug to IT and I can't wait a week for a response. So I type and VOILA the checkbox lights up and I notice there's 3 dots next to it.

Now I have four options, and all four are very helpful yet at the same time take a minute to read through and figure out what each means. Since I do scheduling, the "save and stay" doesn't make a lot of sense to me because I'm all about speed, but I imagine it's helpful for my producer when it's taking an hour to write the short summary of what's happening with Meredith this week and they probably want to save a lot so if it glitches they don't lose their work. Save and new makes sense to me, and discard all changes does too. Then I see the "Save as Copy" -- I'm stoked but also confused because I had to modify my _existing_ record (episode 403) in order to get this "Save as copy" option. I use tons of other software so it's not like I can't think out of the box, but all of them have me duplicate the current record _as is_ before editing, so now I'm not sure if I've changed the title of Episode 403 to jibberish or what so after I click save as copy I go back and check to make sure I didn't overwrite something. I see both items so I'm starting to get it. Only 10000 more times to repeat this til I've got the quarter done. 

The next couple hundred duplications I've got my podcast back on (**_Directus with Ben Haynes_ from "Contributor"**) and I'm making progress -- but since number 1 goal is minimal clicks it's feeling clunking. But I figure there must be a faster way and I'll eventually figure out, so for a bit this becomes my routine:

- click into item
- click into title field
- enter a keystroke, 'abalskdg' (later I realize I'm wasting strokes by not just changing the title to Episode 404 so I finally start doing that)
- click the three dots
- click save as copy
- modify the name field again
- click checkbox to save my "new" episode
- since i'm returned to collection list i find that new one and repeat this process

I don't know yet that all these steps aren't necessary because once I'm "on" the new/duplicate record and made my change I want to immediately duplicate this current duplicate for my next placeholder -- I don't need to go back to the list screen until I finish up with Meredith and move on to _Extreme Makeover: Home Edition._ But I'm not sure if I need to save this copy before I make another copy, or did it save when I clicked save as copy, and sometimes I still doublecheck the record I started with to make sure I didn't mistakenly update that while making a copy because if the Guide says Episode 405 is airing next week and it's supposed to be 404 the fans are going to freak out. So for the next 300 dupes I follow this process:

- click in title field
- update title
- click the three dots
- choose "save and stay" (Now even i have a use for this one! I even learn the shortcode)
- click into title field again
- make a change
- click the three dots
- click "save as copy"
- repeat this list for every dupe 

Eventually I decide there's gotta be a faster way so I do some testing and realize the fastest way is for me to do this:

- click in title field 
- change to what the dupe title needs to be
- click three dots
- click "save as copy" (no shortcode for this one)
- repeat 10,000 times
Honestly, it's not that bad once I figure it out; it's only 4 clicks. Even if I had an icon it would probably still be 4 clicks, maybe it could get down to 3. So my priority 1 does eventually get met, but it took me awhile to get there and since it's not quite as intuitive as what I'm used to I find that I'm just generally slower at this part of my job than I used to be because even when used to it I still sometimes doublecheck I didn't accidentally choose save and stay instead of save as copy and then miss a whole episode. 

So that's a very long use case for a very small thing that seems a bit absurd to be writing as a feature request with so much other stuff to do. But that example has been presented to me dozens of times and is one of the most common questions first asked when we onboard someone to the platform. 

**Overall Conclusion:**

I think it ultimately comes down to the "standard" or most common workflow for duplication in software is to duplicate FIRST "as is" and the word [Copy] appended in some manner, usually to title field, THEN make changes, save, then duplicate again. This is so common that even when the person learns how to do this in Directus they will have used other software in between, thus reinforcing what they know as the "usual" way, so they come back to Directus and often "over click" a few times before they remember the fastest way to do it. 

**Using the drawer idea you suggested, I think if the record could be left unedited AND there was a shortcode for save as copy then it would be fine to keep it in the drawer, leaving the only issue being around does some obvious text like [Copy] get appended or is an in-app notification enough especially as a first approach?**

One final note: Since Directus doesn't want to tell you / restrict you to "this is a title field ALWAYS", I don't have the solution for where to append [Copy] to (what field to use). I'm sure there's many other approaches besides this that could be used, but if this whole long example is about keeping it to one of the "most common" ways this is done here would be my thoughts/questions on how to handle that:

- On the collection settings page in duplication area, allow user to choose / check off a text field to append it to -- next immediate question is well what if the collection only contains booleans, I imagine it's a rabbit hole.....
- Use local cache to hold the text "[Copy]" temporarily and display?
- Temporarily display a message that "record has been duplicated" and as long as it has an autoincrement id allow them to save as that will have increased

**Reply by benhaynes** [reply]
Whew, that's... a response!! Let me try to give some thoughts on the big points.

> A lot of it is absurdly based on number of clicks, so expanding is always that extra click (I DO think this is an awesome idea, my ONLY hesitancy is if the duplicate option goes in there or gets its own icon).

I mentioned the idea of allowing config so you can **choose** which of these menu items would get surfaced as primary action buttons (no menu needed). This could be configured per collection based on the specific workflow needs that exist there. We could explore per-collection+role (or a user preference)... but that might be a big much (not 80/20). Maybe we limit to ~3-4 max primary buttons, and then the rest can be in the context menu. This would allow you to have the key actions be 1-click based on the collection config.

I truly appreciate all your context... would the (hopefully) simple solution I described above streamline the authoring workflow needs you described? Essentially you could choose to have "Duplicate Item" (or any other action) be a primary button. Going further, we could allow for customizing keyboard shortcuts in this same config... so you can really tailor the efficiency, but I don't think that sould be per collection (too confusing to change within a single project).

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
